### PR TITLE
Basic User Auth, Login & Account Creation

### PR DESCRIPTION
**Pull request by jcbang**
- User Login using plaintext email and password post request returns a JWT
- CreateUser API now creates a user in Azure B2C
- APIs secured through Authorization Headers in the format "Bearer token" with manual Issuer, Audience, and Lifespan checking

### CODE REVIEW

File: brainbeats-backend/AuthConnection.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public JwtSecurityToken ValidateToken(string token) {
+      if (!token.StartsWith("Bearer ")) {
+        throw new ArgumentException();
+      }
+
+      token = token.Split(" ")[1];
+
+      string stsDiscoveryEndpoint = "https://ucfbrainbeats.b2clogin.com/ucfbrainbeats.onmicrosoft.com/b2c_1_signupsignin/v2.0/.well-known/openid-configuration";
```
Comment (WarmBuns) [inline brainbeats-backend/AuthConnection.cs]: This should be in a config file and not hardcoded (this url will probably change with Team 2 so lets not make their life harder)

File: brainbeats-backend/Controllers/SampleController.cs [hunk 2]
```diff
@@ -1,22 +1,29 @@
...
+  [ApiController]
+  public class SampleController : ControllerBase {
+    [HttpPost]
+    [Route("create_sample")]
+    public async Task<IActionResult> CreateSample(dynamic req) {
+      try {
+        HttpContext.Request.Headers.TryGetValue("Authorization", out StringValues authorizationToken);
+        AuthConnection.Instance.ValidateToken(authorizationToken);
+      } catch {
+        return BadRequest("Unauthenticated");
```
Comment (WarmBuns) [inline brainbeats-backend/Controllers/SampleController.cs]: What are we catching in all these try/catches? Can there be other errors besides authentication? Shouldnt we be logging these?

File: brainbeats-backend/Controllers/SampleController.cs [hunk 3]
```diff
@@ -26,112 +33,136 @@ public class SampleController : ControllerBase {
         queryString = CreateVertexQuery("sample", sampleId, body, edges);
       } catch {
         return BadRequest("Malformed request");
```
Comment (WarmBuns) [inline brainbeats-backend/Controllers/SampleController.cs]: Would it be possible to return what was malformed about the request?

File: brainbeats-backend/Controllers/UserController.cs [hunk 4]
```diff
@@ -8,13 +11,33 @@ namespace brainbeats_backend.Controllers {
...
   [ApiController]
   public class UserController : ControllerBase {
+    [HttpPost]
+    [Route("login_user")]
+    public async Task<IActionResult> LoginUser(dynamic req) {
+      JObject body = DeserializeRequest(req);
+
+      try {
+        string res = await AuthConnection.Instance.LoginUser(body.GetValue("email").ToString(), 
+          body.GetValue("password").ToString());
```
Comment (WarmBuns) [inline brainbeats-backend/Controllers/UserController.cs:23]: So you're storing the hashed password that the client side hashes right?

Have @hungnguyen1895 and @lloydoad communicated on using the same hash function?

File: brainbeats-backend/Controllers/UserController.cs [hunk 5]
```diff
@@ -8,13 +11,33 @@ namespace brainbeats_backend.Controllers {
...
     public async Task<IActionResult> CreateUser(dynamic req) {
       JObject body = DeserializeRequest(req);
-
       string queryString;
 
+      try {
+        await AuthConnection.Instance.CreateUser($"{body.GetValue("firstName")} {body.GetValue("lastName")}",
+          body.GetValue("email").ToString(), body.GetValue("password").ToString());
+      } catch {
+        return BadRequest("Error in Azure Graph User Creation");
```
Comment (WarmBuns) [inline brainbeats-backend/Controllers/UserController.cs:40]: This is a good error message 👍 the rest should be like this

**Review by WarmBuns — CHANGES_REQUESTED** [review]

File: brainbeats-backend/Controllers/UserController.cs (hunk 4 above)
Comment (jcbang) [reply brainbeats-backend/Controllers/UserController.cs:23]: That's correct

File: brainbeats-backend/Controllers/SampleController.cs (hunk 3 above)
Comment (jcbang) [reply brainbeats-backend/Controllers/SampleController.cs]: Specific errors about what was malformed would require too many case handling for all the possibilities between controllers, and a generic response will need to suffice

File: brainbeats-backend/Controllers/SampleController.cs (hunk 2 above)
Comment (jcbang) [reply brainbeats-backend/Controllers/SampleController.cs]: Updated to return specific error

**Review by WarmBuns — DISMISSED** [review]
LGTM

**Pull request merged**